src/asynchronous/transport.rs

```diff
@@ -0,0 +1,358 @@
+        socket.set_nonblocking(true)?;
+        let socket = tokio::net::UnixStream::from_std(socket)?;
+        Ok(Self(Box::pin(socket)))
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+impl From<tokio_vsock::VsockListener> for Listener {
+    fn from(listener: tokio_vsock::VsockListener) -> Self {
+        let listener = poll_fn(move |cx| -> Poll<Option<IoResult<Socket>>> {
+            let socket = ready!(listener.poll_accept(cx));
+            let socket = socket.map(|(socket, _)| socket.into());
+            Poll::Ready(Some(socket))
+        })
+        .boxed();
+
+        Self(listener)
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+impl From<tokio_vsock::VsockStream> for Socket {
+    fn from(socket: tokio_vsock::VsockStream) -> Self {
+        Self(Box::pin(socket))
+    }
+}
+
+#[cfg(windows)]
+impl TryFrom<tokio::net::windows::named_pipe::NamedPipeServer> for Listener {
+    type Error = IoError;
+    fn try_from(mut listener: tokio::net::windows::named_pipe::NamedPipeServer) -> IoResult<Self> {
+        use std::future::Future;
+
+        let name = get_pipe_name(&listener)?;
+
+        let listener = poll_fn(move |cx| -> Poll<Option<IoResult<Socket>>> {
+            {
+                let fut = std::pin::pin!(listener.connect());
+                if let Err(err) = ready!(fut.poll(cx)) {
+                    return Poll::Ready(Some(Err(err)));
+                };
+            }
+
+            let new_listener =
+                match tokio::net::windows::named_pipe::ServerOptions::new().create(&name) {
+                    Err(err) => return Poll::Ready(Some(Err(err))),
+                    Ok(listener) => listener,
+                };
+
+            let socket = std::mem::replace(&mut listener, new_listener);
+
+            Poll::Ready(Some(Ok(socket.into())))
+        })
+        .boxed();
+
+        Ok(Listener(listener))
+    }
+}
+
+#[cfg(windows)]
+impl From<tokio::net::windows::named_pipe::NamedPipeServer> for Socket {
+    fn from(socket: tokio::net::windows::named_pipe::NamedPipeServer) -> Self {
+        Self(Box::pin(socket))
+    }
+}
+
+#[cfg(windows)]
+impl From<tokio::net::windows::named_pipe::NamedPipeClient> for Socket {
+    fn from(socket: tokio::net::windows::named_pipe::NamedPipeClient) -> Self {
+        Self(Box::pin(socket))
+    }
+}
+
+#[cfg(windows)]
+fn get_pipe_name(pipe: &impl std::os::windows::io::AsRawHandle) -> IoResult<std::ffi::OsString> {
+    use windows_sys::Win32::Storage::FileSystem::*;
+
+    let mut info: FILE_NAME_INFO = unsafe { std::mem::zeroed() };
+    let handle = pipe.as_raw_handle();
+    let success = unsafe {
+        GetFileInformationByHandleEx(
+            handle as _,
+            FileNameInfo,
+            &mut info as *mut FILE_NAME_INFO as _,
+            size_of::<FILE_NAME_INFO>() as _,
+        )
+    };
+    if success != 0 {
+        return Err(IoError::last_os_error());
+    }
```

Comment:
Is this working? This api says:

https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getfileinformationbyhandleex#return-value

```
If the function succeeds, the return value is nonzero and file information data is contained in the buffer pointed to by the lpFileInformation parameter.

If the function fails, the return value is zero. To get extended error information, call [GetLastError](https://learn.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror).
```

Reply:
Good catch!
Fixed